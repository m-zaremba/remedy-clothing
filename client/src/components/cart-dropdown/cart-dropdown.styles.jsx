import styled from 'styled-components';

export const CartDropdownContainer = styled.div`
  position: absolute;
  width: 240px;
  height: 340px;
  display: flex;
  flex-direction: column;
  padding: 20px;
  border: 1px solid black;
  background-color: white;
  top: 90px;
  right: 40px;
  z-index: 5;
`;

export const EmptyMessage = styled.span`
  font-size: 18px;
  margin: auto auto;
  color: black;
`;

export const CartItemsContainer = styled.div`
  height: 240px;
  display: flex;
  flex-direction: column;
  overflow-y: auto;
  scrollbar-color: rgba(80, 80, 80, 0) rgba(0, 0, 0, 0);
  scrollbar-width: thin;
  overflow: auto;
  color: rgba(80, 80, 80, 0);
  transition: all 0.3s;

  &:hover {
    scrollbar-color: rgba(80, 80, 80, 0.4) rgba(0, 0, 0, 0);
    color: rgba(80, 80, 80, 0.45);
  }

  &::-webkit-scrollbar {
    width: 14px;
  }

  &::-webkit-scrollbar-thumb {
    background-clip: content-box;
    border: 4px solid transparent;
    border-radius: 0;
    box-shadow: inset 0 0 0 10px;
  }

  &::-webkit-scrollbar-button {
    width: 0;
    height: 0;
    display: none;
  }

  &::-webkit-scrollbar-corner {
    background-color: transparent;
  }
`;
