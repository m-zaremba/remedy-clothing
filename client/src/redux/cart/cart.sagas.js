import { all, call, put, select, takeLatest } from 'redux-saga/effects';

import { getUserCartRef } from '../../firebase/firebase.utils';
import userActionTypes from '../user/user.types';
import { clearCart, setFirebaseCart } from './cart.actions';
import { selectCurrentUser } from '../user/user.selectors';
import { selectCartItems } from './cart.selectors';
import cartActionTypes from './cart.types';

export function* clearCartOnSignOutOrPaymentSuccess() {
  yield put(clearCart());
}

export function* onPaymentSuccess() {
  yield takeLatest(cartActionTypes.PAYMENT_SUCCESS, clearCartOnSignOutOrPaymentSuccess);
}

export function* onSignOutSuccess() {
  yield takeLatest(userActionTypes.SIGN_OUT_SUCCESS, clearCartOnSignOutOrPaymentSuccess);
}

export function* updateFirebaseCart() {
  const currentUser = yield select(selectCurrentUser);
  if (currentUser) {
    try {
      const cartRef = yield getUserCartRef(currentUser.id);
      const cartItems = yield select(selectCartItems);
      yield cartRef.update({ cartItems });
    } catch (error) {
      console.log(error);
    }
  }
}

export function* checkFirebaseCart({ payload: user }) {
  const cartRef = yield getUserCartRef(user.id);
  const cartSnapshot = yield cartRef.get();
  yield put(setFirebaseCart(cartSnapshot.data().cartItems));
}

export function* onUserSignIn() {
  yield takeLatest(userActionTypes.SIGN_IN_SUCCESS, checkFirebaseCart);
}

export function* onCartChange() {
  yield takeLatest(
    [cartActionTypes.ADD_ITEM, cartActionTypes.REMOVE_ITEM, cartActionTypes.CLEAR_ITEM_FROM_CART],
    updateFirebaseCart,
  );
}

export function* cartSagas() {
  yield all([
    call(onSignOutSuccess),
    call(onPaymentSuccess),
    call(onCartChange),
    call(onUserSignIn),
  ]);
}
