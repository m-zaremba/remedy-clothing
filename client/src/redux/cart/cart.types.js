const cartActionTypes = {
  TOGGLE_CART_HIDDEN: 'TOGGLE_CART_HIDDEN',
  ADD_ITEM: 'ADD_ITEM',
  REMOVE_ITEM: 'REMOVE_ITEM',
  CLEAR_ITEM_FROM_CART: 'CLEAR_ITEM_FROM_CART',
  CLEAR_CART: 'CLEAR_CART',
  PAYMENT_SUCCESS: 'PAYMENT_SUCCESS',
  SET_FIREBASE_CART: 'SET_FIREBASE_CART',
  UPDATE_FIREBASE_CART: 'UPDATE_FIREBASE_CART',
};

export default cartActionTypes;
