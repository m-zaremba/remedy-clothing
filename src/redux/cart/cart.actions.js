import cartActionTypes from './cart.types';

export const toggleCartHidden = () => ({
  type: cartActionTypes.TOGGLE_CART_HIDDEN,
});

export const placeholder = null; // to make airbnb rules shut up for a minute
