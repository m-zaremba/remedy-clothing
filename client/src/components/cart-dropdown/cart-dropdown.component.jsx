import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { withRouter } from 'react-router-dom';

import CartItem from '../cart-item/cart-item.component';

import { selectCartItems } from '../../redux/cart/cart.selectors';
import { toggleCartHidden } from '../../redux/cart/cart.actions';

import {
  CartDropdownContainer,
  EmptyMessage,
  CartItemsContainer,
  CartDropdownButton,
} from './cart-dropdown.styles';

export const CartDropdown = ({ cartItems, history, dispatch }) => {
  return (
    <CartDropdownContainer>
      <CartItemsContainer>
        {cartItems.length ? (
          cartItems.map((cartItem) => <CartItem key={cartItem.id} item={cartItem} />)
        ) : (
          <EmptyMessage>Your cart is empty</EmptyMessage>
        )}
      </CartItemsContainer>
      <CartDropdownButton
        onClick={() => {
          history.push('/checkout');
          dispatch(toggleCartHidden());
        }}
      >
        GO TO CHECKOUT
      </CartDropdownButton>
    </CartDropdownContainer>
  );
};

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems,
});

CartDropdown.propTypes = {
  cartItems: PropTypes.arrayOf(PropTypes.object).isRequired,
  history: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.object, PropTypes.func]),
  ).isRequired,
  dispatch: PropTypes.func.isRequired,
};

export default withRouter(connect(mapStateToProps)(CartDropdown));
