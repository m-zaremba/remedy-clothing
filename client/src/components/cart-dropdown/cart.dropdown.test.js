import React from 'react';
import { shallow } from 'enzyme';

import { CartDropdown } from './cart-dropdown.component';
import CartItem from '../cart-item/cart-item.component';

import { toggleCartHidden } from '../../redux/cart/cart.actions';

describe('testing CartDropdown component', () => {
  let wrapper;
  let mockHistory;
  let mockDispatch;
  const mockCartItems = [{ id: 1 }, { id: 2 }, { id: 3 }];

  beforeEach(() => {
    mockHistory = {
      push: jest.fn(),
    };

    mockDispatch = jest.fn();

    const mockProps = {
      cartItems: mockCartItems,
      history: mockHistory,
      dispatch: mockDispatch,
    };

    wrapper = shallow(<CartDropdown {...mockProps} />);
  });

  it('should render CartDropdown component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call history.push when button is clicked', () => {
    wrapper.find('CartDropdownButton').simulate('click');
    expect(mockHistory.push).toHaveBeenCalled();
    expect(mockDispatch).toHaveBeenCalledWith(toggleCartHidden());
  });

  it('should render an the same number of CartItem components as in the the cartItems prop', () => {
    expect(wrapper.find(CartItem).length).toEqual(mockCartItems.length);
  });

  it('should render EmptyMessage if cartItems is empty', () => {
    const mockProps = {
      cartItems: [],
      history: mockHistory,
      dispatch: mockDispatch,
    };

    const emptyWrapper = shallow(<CartDropdown {...mockProps} />);
    expect(emptyWrapper.exists('EmptyMessage')).toBe(true);
  });
});
