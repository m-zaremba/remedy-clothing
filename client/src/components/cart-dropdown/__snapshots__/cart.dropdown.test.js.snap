// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing CartDropdown component should render CartDropdown component 1`] = `ShallowWrapper {}`;
